src/librustc_typeck/lib.rs

```diff
@@ -138,6 +140,30 @@ pub struct TypeAndSubsts<'tcx> {
     ty: Ty<'tcx>,
 }
 
+fn allow_type_alias_enum_variants<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
+                                                  ty_hir: &hir::Ty,
+                                                  span: Span) -> bool {
+    let allow_feature = tcx.features().type_alias_enum_variants;
+    if !allow_feature {
+        // Only print error if we know the type is an alias.
+        if let hir::TyKind::Path(hir::QPath::Resolved(None, ref path)) = ty_hir.node {
+            if let Def::TyAlias(_) = path.def {
```

Comment:
I don't think this gates `<Enum>::Variant`?

Reply:
Basically, the condition is just `!tcx.features().type_alias_enum_variants` if I'm not missing anything.
If we are getting a variant in `associated_path_def_to_ty` or `resolve_ufcs` then it should be gated.

Reply:
Ah, you're right I think...